### PR TITLE
[SPARK-33964][SQL] Combine distinct unions in more cases

### PR DESCRIPTION
**Pull request by tanelk**
### What changes were proposed in this pull request?

Added the `RemoveNoopOperators` rule to optimization batch `Union`.  Also made sure that the `RemoveNoopOperators` would be idempotent. 

### Why are the changes needed?

In several TPCDS queries the `CombineUnions` rule does not manage to combine unions, because they have noop `Project`s between them.
The `Project`s will be removed by `RemoveNoopOperators`, but by then `ReplaceDistinctWithAggregate` has been applied and there are aggregates between the unions. Adding a copy of `RemoveNoopOperators` earlier in the optimization chain allows `CombineUnions` to work on more queries.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

New UTs and the output of `PlanStabilitySuite`

### CODE REVIEW

File: sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q75.sf100/explain.txt [hunk 1]
```diff
@@ -327,426 +319,386 @@ Join condition: None
...
-(110) Exchange
-Input [7]: [d_year#68, i_brand_id#63, i_class_id#64, i_category_id#65, i_manufact_id#66, sales_cnt#22, sales_amt#23]
-Arguments: hashpartitioning(d_year#68, i_brand_id#63, i_class_id#64, i_category_id#65, i_manufact_id#66, sales_cnt#22, sales_amt#23, 5), true, [id=#74]
-
-(111) HashAggregate [codegen id : 43]
-Input [7]: [d_year#68, i_brand_id#63, i_class_id#64, i_category_id#65, i_manufact_id#66, sales_cnt#22, sales_amt#23]
-Keys [7]: [d_year#68, i_brand_id#63, i_class_id#64, i_category_id#65, i_manufact_id#66, sales_cnt#22, sales_amt#23]
-Functions: []
-Aggregate Attributes: []
-Results [7]: [d_year#68, i_brand_id#63, i_class_id#64, i_category_id#65, i_manufact_id#66, sales_cnt#22, sales_amt#23]
```
Comment (tanelk) [inline sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q75.sf100/explain.txt:630]: All the changes to the plans follow this same 4 node pattern:
- The `Union` is merged with the one above it in the tree.
- The duplicate `Distinct` (aggregate - exchange - aggregate) is removed and its job is done by the one above the merged `Union`

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
cc @cloud-fan

**Comment by HyukjinKwon** [comment]
retest this please

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
retest this please

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master. GA passed.

**Pull request closed without merging**